calour/io.py

```diff
@@ -286,8 +261,48 @@ def read_open_ms(data_file, sample_metadata_file=None, gnps_file=None, feature_m
     return exp
 
 
+def _read_metadata(ids, f, kwargs):
+    '''read metadata table
+
+    Parameters
+    ----------
+    ids : list like of str
+        ids from data table
+    f : str
+        file path of metadata
+    kwargs : dict
+        keyword argument passed to ``pandas.read_table``
+
+    Returns
+    -------
+    ``pandas.DataFrame`` of metadata
+    '''
+    # load the sample/feature metadata file
+    if f is not None:
+        default = {'index_col': 0}
+        if kwargs is None:
+            # use first column as sample ID
+            kwargs = default
+        else:
+            kwargs.update(default)
+        metadata = pd.read_table(f, **kwargs)
+        mid, ids2 = set(metadata.index), set(ids)
+        diff = mid - ids2
+        if diff:
+            logger.warning('These have metadata but do not have data - dropped: %r' % diff)
+        diff = ids2 - mid
+        if diff:
+            logger.warning('These have data but do not have metadata: %r' % diff)
+        # reorder the id in metadata to align with biom
+        metadata = metadata.loc[ids, ]
```

Comment:
what about if some samples/features are not in the metadata file? i think we should add them to the dataframe with "NA" (rather than just showing a warning). Otherwise it will cause bugs downstream.

Reply:
yes. those are added as NA. it is tested in modified tests